docs/3.1/docs/tsa-arch/arch-considerations-upgrading.rst

```diff
@@ -0,0 +1,265 @@
+
+Note that all v1 contracts must be fully upgraded before unvetting v1 to avoid potential issues. This requirement extends beyond a zero-downtime upgrade.
+
+4.2 Testing
+-----------
+Once the Daml packages are vetted, it is essential to ensure that the new version remains compatible with the previous versions. This involves two types of testing:
+
+* A type-level compatibility test checks whether the old and new versions of a package with the same name can coexist without breaking. The easiest way to test this is by uploading both old and new versions to a fresh participant node as part of CI. To do this, access the DAR files used in production. Ideally, these should be stored in a dedicated artifact repository, but given their small sizes (typically under 1 MB), they may also be checked into source control.
+
+* A workflow-level compatibility test verifies that core business processes (workflows) continue to function correctly after an upgrade. At a minimum, it is recommended to include one integration test. A basic integration test should follow these steps:
+
+  1. Start the application with v2 software, but upload only the v1 DAR file to test backward compatibility.
+  2. Initialize the application and start one instance of every core workflow.
+  3. Upload the v2 DAR.
+  4. Update the configuration to instruct the backends to start using the v2 DAR.
+  5. Verify that the workflows remain in the correct state and can continue without issues. 
+
+For more complex upgrades, additional tests may be needed. 
+
+4.3 Rollback
+------------
+
+4.3.1 Rollback by Unvetting
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+To roll back upgrades that do not modify the types of existing templates and choices, unvet the v2 DAR package.
+
+4.3.2 Rollback by Roll-Forward
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+For upgrades that add new fields to existing templates, rollback becomes more complex. In such cases, the rollback must be performed in a “roll-forward” fashion by publishing a new upgrade. This is necessary because if at least one contract has been created using the new fields, those contracts cannot be read with the previous version of the Daml code. Simply unvetting v2 is not an option unless it is acceptable for contracts using the new fields to no longer be referenced. Instead, follow these steps:
+
+1. Publish a new version of the DARs that disregards the newly added fields.
+2. Introduce a `Downgrade` choice in the new version that resets the newly added fields to `None`, making the contracts compatible with the original version.
+3. Use backend automation to iterate through the ACS and invoke the `Downgrade` choice.
+
+To avoid complex "roll-forward" rollbacks, consider breaking an upgrade that introduces new fields into two steps:
+
+1. Introduce an upgrade that adds the new fields but does not use them. Since no changes are made to the choices, this upgrade will not require a rollback in case of a bug.
```

Comment:
```suggestion
1. Introduce an upgrade that adds the new fields but does not use them. Since no changes are made to the choices, this upgrade does not require a rollback in case of a bug.
```